hoomd/mpcd/RandomSubsetPicker.h

```diff
@@ -0,0 +1,292 @@
+        GPUArray<unsigned int> m_flags_idx;                  //!< For storing index of flags which are 1 
+
+        //! For calculating total number of particles which has 1 in flags[0 1 0 1 1 0...] and storing their index in flags_idx
+        template<typename T>
+        unsigned int CalculateTotalNumbersToPickFrom(const GPUArray<T>& flags, unsigned int N);
+
+        //!< For Making a random pick of particles across all ranks
+        void makeAllPicks(unsigned int timestep, unsigned int N_try_pick, unsigned int N_total_all_ranks);
+        //!< For storing indices of picked particles in picks
+        void storePicksIdx(GPUArray<unsigned int>& picks);
+
+    private:
+        std::vector<unsigned int> m_all_picks;               //!< All picked particles on all the ranks
+        //!< Indices of Particles picked on this rank in a /m_flags_idx array( This will not contain the original index of picked particles in flags, these are index of particles picked in m_flags_idx))
+        GPUArray<unsigned int> m_picks_idx;
+    };
+
+/*
+ * \param picks       //!< indexes that are picked in the *original* flags array
+ * \param N_pick      //!< number of particles picked
+ * \param flags       //!< flags indicating which particles to pick from [ 0 1 0 1 1 ...] 
+ * \param N_try_pick  //!< target number of particles to pick (if total number of 1's in flags will be less than N_try_pick, it will pick all the 1's, in that case N_pick would be less then N_try_pick
+ * \param seed        //!< Seed to random number Generator
+ * \param timestep    //!< timestep at which this is called
+ * \param N           //!< total number of particles 
+ */
+template<typename T>
+void RandomSubsetPicker::operator()(GPUArray<unsigned int>& picks,
+                                    unsigned int& N_pick,
+                                    const GPUArray<T>& flags,
+                                    unsigned int N_try_pick,
+                                    unsigned int timestep,
+                                    unsigned int N)
+    {
+    /*
+     * It first calculates total number of 1's in flags(N_total) and store the indices of those 1's in m_flags_idx.
+     * then it calculates total number of 1's on all the ranks(N_total_all_ranks).
+     * If N_total_all_ranks < N_try_pick it picks all the particles 
+     * else it does more complicated picking logic
+     *
+     *
+     * In the end it stores the indexes of picked particles(indices of picked particles according to the *original* flags array.) in picks and 
+     * number of particles picked to N_pick indexes that were picked in the *original* flags array.
+     */
+
+    /*
+     * First calculating How many marked particles are there in flags(1 = marked) and storing their indexes in m_flags_idx
+     * This is done by CalculateTotalNumbersToPickFrom()
+     */
+    unsigned int N_total = CalculateTotalNumbersToPickFrom<T>(flags, N);
```

Comment:
I think this function call actually does a lot more than what its name implies, so I would try to think of a better name. Maybe like `countAndCompactFlags`?